### ISSUE TITLE
@UnstableApi classes from Netty are used causing errors with 4.1.36.Final

### ISSUE THREAD

**Issue by adrian-baker** [aws/aws-sdk-java-v2#1344]
The SDK uses classes marked as `@UnstableApi` from Netty. As a result, minor version upgrades to Netty can break the SDK.

For example, when running 2.7.4 with 4.1.36.Final, I'm seeing this exception, which I believe is caused by the use of an unstable API changing between 4.1.33.Final and 4.1.36.Final.

```
19:29:01.408 - [aws-java-sdk-NettyEventLoop-1-1] WARN  io.netty.channel.ChannelInitializer - Failed to initialize a channel. Closing: [id: 0xa04b5a4b]
java.lang.IllegalStateException: failed to build a Http2ConnectionHandler
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodecBuilder.buildFromCodec(ForkedHttp2MultiplexCodecBuilder.java:189)
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodecBuilder.buildFromConnection(ForkedHttp2MultiplexCodecBuilder.java:179)
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodecBuilder.build(ForkedHttp2MultiplexCodecBuilder.java:153)
        at software.amazon.awssdk.http.nio.netty.internal.ChannelPipelineInitializer.configureHttp2(ChannelPipelineInitializer.java:125)
        at software.amazon.awssdk.http.nio.netty.internal.ChannelPipelineInitializer.channelCreated(ChannelPipelineInitializer.java:85)
        at io.netty.channel.pool.SimpleChannelPool$1.initChannel(SimpleChannelPool.java:113)
        at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)
        at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:964)
        at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:610)
        at io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:46)
        at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1461)
        at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1126)
        at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:651)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:515)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:487)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:405)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:906)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoSuchMethodError: io.netty.handler.codec.http2.Http2FrameCodec.<init>(Lio/netty/handler/codec/http2/Http2ConnectionEncoder;Lio/netty/handler/codec/http2/Http2ConnectionDecoder;Lio/netty/handler/codec/http2/Http2Settings;)V
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodec.<init>(ForkedHttp2MultiplexCodec.java:171)
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodecBuilder.build(ForkedHttp2MultiplexCodecBuilder.java:203)
        at io.netty.handler.codec.http2.ForkedHttp2MultiplexCodecBuilder.buildFromCodec(ForkedHttp2MultiplexCodecBuilder.java:185)
```

A similar exception was reported against Netty in https://github.com/netty/netty/issues/7854, which was closed with :

> Http2ConnectionEncoder and Http2FrameWriter are flagged as @UnstableApi so breaking changes can happen anytime.

To workaround, I had to hunt down the version of netty specified in the pom, and force the downgrade in my gradle project:

`implementation enforcedPlatform("io.netty:netty-bom:4.1.33.Final")`

**Comment by dagnir** [comment]
Hi @adrian-baker, apologies for the inconvenience. Unfortunately, Netty's HTTP/2 APIs are still in the processing of being stabilized (https://github.com/netty/netty/issues/8185) so there's no good solution for this for now except for the SDK to make changes as we upgrade the Netty dependency.

**Comment by brharrington** [comment]
@dagnir have you considered shading the netty dependency to avoid conflicts with other usage? gRPC ended up doing that a while back for similar reasons (https://github.com/grpc/grpc-java/pull/2485).

**Comment by dagnir** [comment]
Ah cool, thanks @brharrington for the pointer! Looks like they've solved no ungradability issue by providing a shaded and non-shaded version. I'll discuss our options with the team and we'll see if this is a viable solution for us.

**Comment by zoewangg** [comment]
The fix has been released as part of `2.8.0`. Closing the issue. Feel free to reopen if you have further questions.
